sw/airborne/modules/sensors/airspeed_sdp3x.c

```diff
@@ -0,0 +1,223 @@
+ * default airspeed scale is 2/1.225
+ */
+#ifndef SDP3X_AIRSPEED_SCALE
+#define SDP3X_AIRSPEED_SCALE 1.6327
+#endif
+
+struct AirspeedSdp3x sdp3x;
+static struct i2c_transaction sdp3x_trans;
+
+static bool sdp3x_crc(const uint8_t data[], unsigned size, uint8_t checksum)
+{
+  uint8_t crc_value = 0xff;
+
+  // calculate 8-bit checksum with polynomial 0x31 (x^8 + x^5 + x^4 + 1)
+  for (unsigned i = 0; i < size; i++) {
+    crc_value ^= (data[i]);
+
+    for (int bit = 8; bit > 0; --bit) {
+      if (crc_value & 0x80) {
+        crc_value = (crc_value << 1) ^ 0x31;
+
+      } else {
+        crc_value = (crc_value << 1);
+      }
+    }
+  }
+
+  // verify checksum
+  return (crc_value == checksum);
+}
+
+static void sdp3x_downlink(struct transport_tx *trans, struct link_device *dev)
+{
+  int16_t temp = (int16_t)(sdp3x.temperature * 10.f);
+  pprz_msg_send_AIRSPEED_MS45XX(trans,dev,AC_ID,
+                                &sdp3x.pressure,
+                                &temp,
+                                &sdp3x.airspeed);
+}
+
+void sdp3x_init(void)
+{
+  sdp3x.pressure = 0.;
+  sdp3x.temperature = 0;
+  sdp3x.airspeed = 0.;
+  sdp3x.pressure_scale = SDP3X_PRESSURE_SCALE;
+  sdp3x.pressure_offset = SDP3X_PRESSURE_OFFSET;
+  sdp3x.airspeed_scale = SDP3X_AIRSPEED_SCALE;
+  sdp3x.sync_send = SDP3X_SYNC_SEND;
+  sdp3x.initialized = false;
+
+  sdp3x_trans.status = I2CTransDone;
+  // setup low pass filter with time constant and 100Hz sampling freq
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_AIRSPEED_MS45XX, sdp3x_downlink); // FIXME
+#endif
+}
+
+void sdp3x_periodic(void)
+{
+  if (sdp3x_trans.status != I2CTransDone) {
+    return; // not ready
+  }
+
+  if (sdp3x.initialized) {
+    // Initiate next read
+    i2c_receive(&SDP3X_I2C_DEV, &sdp3x_trans, SDP3X_I2C_ADDR, 6);
+  }
+  else {
+    // Init sensor in continuous mode
+    sdp3x_trans.buf[0] = SDP3X_MODE >> 8;
+    sdp3x_trans.buf[1] = SDP3X_MODE & 0xFF;
+    i2c_transmit(&SDP3X_I2C_DEV, &sdp3x_trans, SDP3X_I2C_ADDR, 2);
+    sdp3x.initialized = true;
+  }
+}
+
+void sdp3x_event(void)
+{
+  /* Check if transaction is succesfull */
+  if (sdp3x_trans.status == I2CTransSuccess) {
+
+    if (sdp3x.initialized) {
+      // make local copy of buffer
+      uint8_t buf[6];
+      for (uint8_t i = 0; i < 6; i++) {
+        buf[i] = sdp3x_trans.buf[i];
+      }
+
+      // Check the CRC
+      if (!sdp3x_crc(&buf[0], 2, buf[2]) || !sdp3x_crc(&buf[3], 2, buf[5])) {
+        // error
+        sdp3x_trans.status = I2CTransDone;
+        return;
+      }
+
+      int16_t p_raw = ((int16_t)(buf[0]) << 8) | (int16_t)(buf[1]);
+      sdp3x.pressure = ((float)p_raw / sdp3x.pressure_scale) - sdp3x.pressure_offset;
+
+      int16_t t_raw = ((int16_t)(buf[3]) << 8) | (int16_t)(buf[4]);
+      sdp3x.temperature = (float)t_raw / SDP3X_SCALE_TEMPERATURE;
+
+      // Send (differential) pressure via ABI
+      AbiSendMsgBARO_DIFF(SDP3X_SENDER_ID, sdp3x.pressure);
```

Comment:
Adding the AbiSendMsgBARO_DIFF and AbiSendMsgTEMPERATURE inside the next "#if USE_AIRSPEED_SPD3X" condition lets us use the sensor as an external sensor when required. For example, we may use the MS45XX airspeed sensor for flight which updates the state and use the SDP3x for other purpose ?

Reply:
If you have several sensor sending the same type of data but you only want to have one of them used, you should filter from the receiver side, so not bind to ABI_BROADCAST but to the specific sensor

Reply:
Ok!, I get it now...